### DISCUSSION TITLE
Calling ItemsService inside a "items.create.before" hook blocks the server

### DISCUSSION THREAD

**Discussion by laperla** [directus/directus#6888] [Archive]
Hi everyone,

I'm using Directus for the first time and I think I'm missing something about database blocking and transactions. I was deploying a hook that needs to query some collection for checking some validation constraint before creating the item. But when I call to ItemsService, the server gets fully blocked and doesn't respond aymore. I tested the same query code inside a "items.created" hook, and executes correctly. 

hope someone can give me some tip, I'm simply trying to validate input by querying other collections.

Thanks!!

**Comment by rijkvanzanten** [comment]
What database vendor are you using, and can you share a copy of your hook? Hard to tell what's going wrong without :)

**Reply by laperla** [reply]
oh sorry, i totally forgot it, I'm using a SQlite3 database and my hook look like this

```js
module.exports = function registerHook({ services, exceptions }) {

    const { ItemsService } = services

    return {
        "items.create.before": async (input, { collection, schema, accountability }) => {

            // Just for testing the behaivour
            const clientsService = new ItemsService("clients", { schema, accountability })
            await clientsService.readByQuery({ fields: ['*'] }).then(console.log)

            return input
        },
    }
}
```

**Reply by rijkvanzanten** [reply]
No worries 🙂 

SQLite can only do a single transaction at a time, so your call to `readByQuery` is a second query that's opened during the parent transaction. You can hook into the parent transaction by using the transaction database instance in the service

Can you try the following?

```js
module.exports = function registerHook({ services, exceptions }) {

    const { ItemsService } = services

    return {
        "items.create.before": async (input, { collection, schema, accountability, database }) => {

            // Just for testing the behaivour
            const clientsService = new ItemsService("clients", { schema, accountability, knex: database })
            await clientsService.readByQuery({ fields: ['*'] }).then(console.log)

            return input
        },
    }
}
```

@nickrum can we find a good place to document this?

**Reply by laperla** [reply]
yay it works!! :)

Previously I dig into the AbstractServiceOptions interface and reached to the same solution, but I applyed it wrongly:

`
 // 'database' isn't assigned to 'knex' property (and then nested transaction)
const clientsService = new ItemsService("clients", { schema, accountability, database }) 
`

Thanks a lot!

**Reply by nickrum** [reply]
@rijkvanzanten Having an `ItemsService` example inside the custom hook docs would be nice in general. We could just add it there as the default. Passing the database instance shouldn't hurt when using database vendors other than SQLite.

**Reply by MatanMaimon** [reply]
> No worries 🙂
> 
> SQLite can only do a single transaction at a time, so your call to `readByQuery` is a second query that's opened during the parent transaction. You can hook into the parent transaction by using the transaction database instance in the service
> 
> Can you try the following?
> 
> ```ts
> module.exports = function registerHook({ services, exceptions }) {
> 
>     const { ItemsService } = services
> 
>     return {
>         "items.create.before": async (input, { collection, schema, accountability, database }) => {
> 
>             // Just for testing the behaivour
>             const clientsService = new ItemsService("clients", { schema, accountability, knex: database })
>             await clientsService.readByQuery({ fields: ['*'] }).then(console.log)
> 
>             return input
>         },
>     }
> }
> ```
> 
> @nickrum can we find a good place to document this?

Hey
Can you help why this code works while this one isn't:
```js
import { defineHook } from "@directus/extensions-sdk";
import {
  ForbiddenError,
  createError,
  InternalServerError,
} from "@directus/errors";

interface MapPayload {
  event: string | { id: string };

  // ...
}

const collectionName = "maps";

export default defineHook(({ filter }, { services, getSchema, database }) => {
  const { ItemsService } = services;

  filter(
    "items.create",
    async (
      payload: MapPayload,
      meta: Record<string, any>,
      context: { accountability: any }
    ) => {
      if (meta.collection !== collectionName) {
        return payload;
      }

      const userId = context.accountability?.user;
      if (!userId) {
        throw new ForbiddenError();
      }

      const relationFieldCollectionName = "events";
      const relationFieldIdValue =
        typeof payload?.event === "string"
          ? payload?.event
          : payload?.event?.id;

      const schema = await getSchema();
      const eventService = new ItemsService(relationFieldCollectionName, {
        schema,
        accountability: context.accountability,
        knex: database,
      });

      const event = await eventService.readOne(relationFieldIdValue, {
        fields: ["*"],
      });

      console.log({ event });

      return payload;
    }
  );
});
```

after waiting a lot of time for request, it's gets the same `knex timeout` error even if I use `knex: database`.
using `.then` without awaiting to `readOne` works but it won't help me (my real use-case is to await to check nested relation field permission)...

**Reply by MatanMaimon** [reply]
Just found that when using `database` (`knex`) from `context` it works! (saw [this ](https://github.com/directus/directus/discussions/18050#discussioncomment-5932737) and trying that approach).
but why? what's the difference?

anyway, my use case was to check if the user creating the `map`, is trying to connect it to the `event` he's not the owner of.

if it will help to someone, this is the working code:
```ts
import { defineHook } from "@directus/extensions-sdk";
import { ForbiddenError } from "@directus/errors";

interface MapPayload {
  event: string | { id: string };
  // ...
}

const relationFieldCollectionName = "events";

export default defineHook(({ filter }, { services }) => {
  const { ItemsService } = services;

  filter(
    "maps.items.create",
    async (
      payload: MapPayload,
      _,
      context: { accountability: any; database: any; schema: any }
    ) => {
      const userId = context.accountability?.user;
      if (!userId) {
        throw new ForbiddenError();
      }

      const relationFieldIdValue =
        typeof payload?.event === "string"
          ? payload?.event
          : payload?.event?.id;

      const { schema, accountability, database: knex } = context;

      const eventService = new ItemsService(relationFieldCollectionName, {
        schema,
        accountability,
        knex,
      });

      const event = await eventService.readOne(relationFieldIdValue, {
        filter: {
          user_created: {
            _eq: userId,
          },
        },
      });

      if (!event) {
        throw new ForbiddenError();
      }

      return payload;
    }
  );
});

```

**Comment by hmedhappy** [comment]
Hey Guys !
I Am new am new in directus 
i want to update a variable in a collections 
- i created a cron Jon and linked it with a script 
but the data dosen't  change at all 
anyone can tell me why ? here's what i did 

in the script operation ( linked with the job )

```js
module.exports = function ({ action }, { services, exceptions }) {
    const { ItemsService } = services;
  
    // Send mail to admin when event time is updated
    action('items.update', async ({ collection, schema, accountability, database }) => {
    
	const configService = new ItemsService("Config", { schema, accountability, knex: database })


      await configService.updateOne("6446a7bd-7c04-4e3e-8218-49e6c1d511c2",{test :70 })
        
  
      return  ;
    });
  };
```

PS : The job DOES work every minute but it doesn't affect the data

**Reply by azrikahar** [reply]
I believe there might be a confusion here, as the code you shared seems to be for [hooks extensions](https://docs.directus.io/extensions/hooks.html), which does not work as-is by placing it in Flows' Run Script operation. 

Assuming the cron job you mentioned _is_ a Flows [Schedule trigger](https://docs.directus.io/configuration/flows/triggers.html#schedule), then you should use the [Update Data operation](https://docs.directus.io/configuration/flows/operations.html#update-data) to perform the update, rather than a Run Script operation 👍 (you can of course still use the Run Script operation to do some calculation which can be used in the Update Data operation's Payload)

**Reply by hmedhappy** [reply]
Yes Exactly : 
All i want to do is to increment the current week number ( 1-2-3-4 ) , i can't do by a simple payload cauz i need a pre-calculation before in a Run Script BUT THE PROBLEM IS ,AM NOT SURE IF I NEED ( action - filter - init - schedule ) or none of them .
All i need is : 
- each time the crone get triggered 
- activate Run Script to calculate et prepare the payload 
- Make a Update Data operation with the payload 
THATS IT .
Here's my Flow Design :
![image](https://user-images.githubusercontent.com/48823914/189665227-9ceac887-e322-4fa9-be0a-034986124f76.png) ;

and here's my run script:

```js
module.exports = function registerHook({ action }, { services, exceptions }) {
    const { ConfigService } = services;
  
    // Send mail to admin when event time is updated
    action('config.update', async ({ keys, collection, payload }, { schema, accountability }) => {
      if (collection !== 'events') return;
    
  
      const configService = new ConfigService('Config', { accountability: accountability, schema: schema });
      const config = await configService.readOne("6446a7bd-7c04-4e3e-8218-49e6c1d511c2");

        
      const curren_week = config.week ;
      const updated_week = curren_week === 4 ? 1 : curren_week + 1
        
  
      return await configService.updateOne("6446a7bd-7c04-4e3e-8218-49e6c1d511c2",{week :updated_week })
    });
  };
```

Do i need a operation such us action or schedule or filter for this type of update ?
if not , how can i write a Run Script that check the current value from the config table and then update it with a new calculated variable

**Reply by rijkvanzanten** [reply]
@hmedhappy You're mixing concepts here. You can either achieve this through a flow, or achieve this through a hook. No need to do both at the same time. 

In your code example above, you're using the `'config.update'` `action` hook rather than a cron job, which might explain why it's not running when you expect it is
